File: backend/roia/ranking.py
"""The ranking arithmetic — pure Python, no LLM.

The overall score is a weighted mean the browser can recompute with the identical formula,
which is the only reason the weight sliders can be honest (AC6). A model that produced the
overall directly would make the sliders decorative and the report unreproducible.

``competitive_differentiation`` and ``collaboration_potential`` are ``None`` in the demo
(non-goal §4): excluded from the sum, rendered "not assessed", and the remaining weights
renormalised over the scored seven.
"""

from __future__ import annotations

import math
from collections.abc import Mapping
from typing import Literal

from roia.llm_schemas import CRITERIA, CriterionScore

#: Flat by default. Grant-derived weights are `plan.md` §7.3 and out of scope here.
#: All nine keys ship to the browser so every slider has something to move, including the
#: two whose criterion is unscored — moving those changes nothing, which is the truth.
DEFAULT_WEIGHTS: dict[str, float] = {name: 1 / len(CRITERIA) for name in CRITERIA}

#: `fixtures/report-sample.json` stores `overall` to 2 dp. Round the same way or a golden
#: test fails on float noise that means nothing.
PLACES = 2
SCALE = 10**PLACES

#: The `evidence_strength` rubric's own anchors, read off `llm.py`'s RUBRICS — **not numbers
#: chosen to taste**:
#:   8 — "several papers' abstracts were retrieved, and a trend and a citing count both
#:       point the same way"
#:   5 — "several retrieved records support the claim, but none were read beyond their titles"
#:   2 — "the claim rests on one retrieved record, or on a query that returned almost nothing"
#: So `high` is the 8-anchor, `medium` the 5-anchor, and everything below is `low`.
CONFIDENCE_HIGH = 8.0
CONFIDENCE_MEDIUM = 5.0

Confidence = Literal["high", "medium", "low"]


def confidence_for(
    scores: Mapping[str, CriterionScore | None], *, thin_evidence: bool = False
) -> Confidence:
    """How much the evidence supports the direction — **derived, never asserted.**

    This used to be a field the model filled in about its own output, which is the one thing
    it is least able to judge and the only place in the report where a claim had nothing
    behind it. Everything else here is checkable: each score cites evidence, the overall is
    arithmetic the browser repeats. Now confidence is too — it is a reading of
    `evidence_strength`, which is itself scored against a rubric and carries its own
    citations.

    On the two recorded runs this reproduces what the model said for all six directions, so
    the change buys a guarantee rather than different answers.

    ``thin_evidence`` caps it at ``low``. That flag means the direction ended up below the
    citation floor after unresolvable ids were dropped, which is a fact about the direction
    the per-criterion score cannot see — a model can score `evidence_strength` 8 on evidence
    that later fails to resolve.
    """
    strength = scores.get("evidence_strength")
    if strength is None or thin_evidence:
        return "low"
    if strength.value >= CONFIDENCE_HIGH:
        return "high"
    if strength.value >= CONFIDENCE_MEDIUM:
        return "medium"
    return "low"


def _round_half_up(value: float) -> float:
    """Round to `PLACES` the way the browser does — bit for bit, not merely similarly.

    `ranking.ts` computes `Math.round(v * 100) / 100`, which is half-**up**. The builtin
    `round()` is half-to-**even**, so the two copies disagreed on every value landing exactly
    on a half: a raw 5.625 was 5.62 here and 5.63 in the browser. Against the real `run-004`
    report that is 1.2% of slider positions — the backend writing one number into
    `report.json` while the slider beside it shows another, which is the drift AC6 exists to
    rule out.

    `math.floor(v * SCALE + 0.5)` is what `Math.round` is *defined* as, evaluated in the same
    order over the same IEEE-754 doubles, so it agrees with the browser on every input rather
    than on most of them. `Decimal.quantize(ROUND_HALF_UP)` would round the true decimal value
    rather than the double and put the two back out of step wherever `v * SCALE` is inexact.

    Nothing already written moves: at the default flat weights all six recorded runs round
    identically either way. This changes only the numbers the sliders produce, and changes
    them to the ones already on screen.
    """
    return math.floor(value * SCALE + 0.5) / SCALE


def compute_ranking(
    scores: Mapping[str, CriterionScore | None],
    weights: Mapping[str, float] = DEFAULT_WEIGHTS,
) -> float:
    """The weighted mean over the criteria that were actually scored.

    Renormalising over the scored subset is what makes flat weights give the plain mean of
    the seven, rather than five-ninths of it. A direction with nothing scored is 0.0, not a
    division by zero.
    """
    #: Accumulated **exactly** as `ranking.ts` accumulates it: sum `weight * value`, then
    #: divide once at the end. Dividing each term by `total` first is algebraically the same
    #: and numerically is not — it diverged from the browser on 0.08% of weight vectors even
    #: after the rounding was fixed, because the two sides were summing different floats
    #: before they ever reached the rounding. Same operations, same order, same doubles.
    total = 0.0
    weighted = 0.0
    for name, score in scores.items():
        weight = weights.get(name, 0.0)
        if score is None or weight <= 0:
            continue
        total += weight
        weighted += weight * score.value
    if total <= 0:
        return 0.0
    return _round_half_up(weighted / total)

File: frontend/src/ranking.ts
/**
 * The ranking arithmetic, mirrored from `backend/roia/ranking.py`.
 *
 * This is the one piece of logic that deliberately exists twice, and it has to agree to the
 * last decimal: AC6 says a slider re-orders the directions **with zero network requests**,
 * so the browser recomputes rather than asks. A test asserts that at the default weights
 * this reproduces the `overall` the backend already wrote into the report — if the two ever
 * drift, that fails rather than the sliders quietly lying.
 *
 * Rule 2 of the project still holds: **no model is involved in either copy.**
 */

import type { CriterionScore } from './types'

/** `llm_schemas.CRITERIA`, in §6.4 order. The matrix and the sliders both render in this order. */
export const CRITERIA = [
  'grant_alignment',
  'scientific_novelty',
  'importance',
  'applicant_fit',
  'feasibility',
  'competitive_differentiation',
  'collaboration_potential',
  'impact_potential',
  'evidence_strength',
] as const

export type Criterion = (typeof CRITERIA)[number]

/**
 * Scored `null` in the demo — the competitor and collaborator stage is a non-goal (§4).
 * They still get a slider, because moving it changing nothing is the honest thing to show.
 */
export const NOT_ASSESSED: readonly Criterion[] = [
  'competitive_differentiation',
  'collaboration_potential',
]

/** Sliders are integers so the URL stays exact — no float noise round-tripping through it. */
export const WEIGHT_MIN = 0
export const WEIGHT_MAX = 10
export const WEIGHT_DEFAULT = 5

export type Weights = Record<string, number>

export const DEFAULT_WEIGHTS: Weights = Object.fromEntries(
  CRITERIA.map((name) => [name, WEIGHT_DEFAULT]),
)

export function label(criterion: string): string {
  return criterion.replace(/_/g, ' ').replace(/^./, (c) => c.toUpperCase())
}

/**
 * The weighted mean over the criteria that were actually scored — `compute_ranking()` in
 * `ranking.py`, line for line.
 *
 * Renormalising over the scored subset is what makes equal weights give the plain mean of
 * the seven rather than five-ninths of it, and it is why integer sliders reproduce the
 * backend's flat `1/9` weights exactly: 5/35 and (1/9)/(7/9) are both 1/7.
 */
export function computeOverall(
  scores: Record<string, CriterionScore | null>,
  weights: Weights = DEFAULT_WEIGHTS,
): number {
  let total = 0
  let weighted = 0
  for (const [name, score] of Object.entries(scores)) {
    const weight = weights[name] ?? 0
    if (score === null || weight <= 0) continue
    total += weight
    weighted += weight * score.value
  }
  if (total <= 0) return 0
  // `compute_ranking` in ranking.py performs these same operations in this same order, on
  // purpose. Two things had to be fixed to make that true, and both are easy to undo:
  //   1. it rounded with Python's builtin `round()` (half-to-even), so a raw 5.625 became
  //      5.62 there and 5.63 here;
  //   2. it divided each term by `total` before summing instead of dividing once at the end,
  //      which is algebraically the same and numerically is not — that alone left 0.08% of
  //      weight vectors disagreeing after the rounding was fixed.
  // So: do not "simplify" the accumulation below, and if you change the rounding, change it
  // there in the same commit. A sweep test in test_report.py fails if these drift apart.
  return Math.round((weighted / total) * 100) / 100
}

export interface Ranked<T> {
  item: T
  overall: number
  rank: number
}

/**
 * Re-rank by the recomputed overall. Ties keep their original order, so a slider that
 * changes nothing does not shuffle the cards.
 */
export function rankBy<T extends { scores: Record<string, CriterionScore | null> }>(
  items: readonly T[],
  weights: Weights,
): Ranked<T>[] {
  return items
    .map((item, index) => ({ item, overall: computeOverall(item.scores, weights), index }))
    .sort((a, b) => b.overall - a.overall || a.index - b.index)
    .map((entry, position) => ({ item: entry.item, overall: entry.overall, rank: position + 1 }))
}

// --- weights in the URL ---------------------------------------------------------------------

/**
 * Nine integers in `CRITERIA` order, e.g. `?w=5,5,9,5,5,5,5,5,5`. Positional rather than
 * named so the URL stays short enough to paste into a chat window, which is the whole point
 * of putting the weights there: "look at it with novelty turned up" is one link.
 */
export function encodeWeights(weights: Weights): string {
  return CRITERIA.map((name) => weights[name] ?? WEIGHT_DEFAULT).join(',')
}

export function decodeWeights(encoded: string | null): Weights {
  if (!encoded) return DEFAULT_WEIGHTS
  const parts = encoded.split(',')
  if (parts.length !== CRITERIA.length) return DEFAULT_WEIGHTS
  const weights: Weights = {}
  for (const [index, name] of CRITERIA.entries()) {
    const value = Number(parts[index])
    // A hand-edited URL must not be able to produce NaN weights and a blank report.
    weights[name] = Number.isFinite(value)
      ? Math.min(WEIGHT_MAX, Math.max(WEIGHT_MIN, Math.round(value)))
      : WEIGHT_DEFAULT
  }
  return weights
}

export function isDefault(weights: Weights): boolean {
  return CRITERIA.every((name) => (weights[name] ?? WEIGHT_DEFAULT) === WEIGHT_DEFAULT)
}
